proposed/http-handlers/request-handlers-meta.md

```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+
+This approach has the following benefits:
+
+- Middleware does not need to know anything about any other middleware or how it
+  is composed in the application.
+- The `QueueRequestHandler` is agnostic of the PSR-7 implementation in use
+- Middleware is executed in the order it is added to the application, making the
+  code explicit.
+
+#### Decoration-based request handler
+
+In this approach, a request handler implementation decorates both a middleware
+instance and a fallback request handler to pass to it. The application is built
+from the outside-in, passing each request handler "layer" to the next outer one.
+
+```php
+class DecoratedRequestHandler implements RequestHandlerInterface
+{
+    private $middleware;
+    private $nextHandler;
+
+    public function __construct(MiddlewareInterface $middleware, RequestHandlerInterface $nextHandler)
+    {
+        $this->middleware = $middleware;
+        $this->nextHandler = $nextHandler;
+    }
+
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        return $this->middleware->process($request, $this->nextHandler);
+    }
+}
+
+$innerHandler = new class ($responsePrototype) implements RequestHandlerInterface {
+    private $responsePrototype;
+
+    public function __construct(ResponseInterface $responsePrototype)
+    {
+        $this->responsePrototype = $responsePrototype;
+    }
+
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        return $this->responsePrototype;
+    }
+};
+
+$layer1 = new DecoratedRequestHandler(new RoutingMiddleware(), $innerHandler);
+$layer2 = new DecoratedRequestHandler(new CorsMiddleware(), $layer1);
+
+$response = $layer2->handle(ServerRequestFactory::fromGlobals());
+```
+
+Similar to the queue-based middleware, request handlers serve two purposes in
+this system:
+
+- Producing a fallback response if no other layer does.
+- Dispatching middleware.
+
+#### Reusable Middleware Examples
+
+In the examples above, we have two middleware composed in each. In order for
+these to work in either situation, we need to write them such that they interact
+appropriately.
+
+Generally speaking, implementors of middleware should follow these guidelines:
+
+- Test the request for a required condition. If it does not satisfy that
+  condition, use a composed prototype response or a composed response factory
+  to generate and return a response.
+
+- If pre-conditions are met, delegate creation of the response to the provided
+  request handler, optionally providing a "new" request by manipulating the
+  provided request (e.g., `$handler->handle($request->withAttribute('foo',
+  'bar')`).
+
+- Either pass the response returned by the request handler unaltered, or provide
+  a new response by manipulating the one returned (e.g., `return
+  $response->withHeader('X-Foo-Bar', 'baz')`).
+
+The `CorsMiddleware` is one that will exercise all three of these guidelines: if
```

Comment:
I am mildly uncomfortable using a highly specific example like this. This assumes that I know what CORS is and what the failure conditions mean. If we're going to use a specific example perhaps something like routing would be more generally understood?

Reply:
I'll see if I can come up with another example. This was one I liked because it hits all the various ways middleware might produce a response:

- on its own, due to a failed pre-condition
- by acting as a pass-through to the request handler
- by modifying the response returned by the request handler

Let me see if I can come up with an example, and I'll ping you if/when I do.

Reply:
Came up with an example using authorization instead of CORS; please review and let me know what you think.